README.md
# [IRChuu](https://26000.github.io/irchuu/) [![](https://goreportcard.com/badge/github.com/26000/irchuu)](https://goreportcard.com/report/github.com/26000/irchuu)
![IRChuu in WeeChat](https://26000.github.io/irchuu/images/shadow/Screenshot_20170319_092103.png)

A Telegram<->IRC transport.

## Features
- Relays messages between a Telegram (super)group and an IRC channel
- Lightweight, written in Go. Consumes only around 10MiB RAM!
- IRC authentication using SASL or NickServ
- (optional) Keeps log of the chat in a PostgreSQL database (those who recently joined the IRC channel can view history!)
- Preserves markup: bold in Telegram will remain bold in IRC
- All Telegram media types support; serves or uploads files so they are accessible in IRC
- All Telegram features like forwards, replies and edits are also supported
- Coloured nicknames in IRC
- (optional) Telegram group administrators can moderate the IRC channel and vice versa
- ...and this is not a complete list!

## Setup
### Installation
> **Note** IRChuu is written in Go and should work on all platforms supported by Go, including Linux, macOS and Windows. Nevertheless, I strongly recommend you using a Linux flavour to run IRChuu, or, as a last resort, macOS. IRChuu is only tested on Linux.

You need to install [golang](https://golang.org/doc/install), [git](https://git-scm.com/downloads) and configure your `$GOPATH`. Just set `$GOPATH` environment variable to a writable directory and add `$GOPATH/bin` to your system `$PATH`.

After that, install IRChuu~:
```
$ go get github.com/26000/irchuu/...
```
Upgrade with:
```
$ go get -u github.com/26000/irchuu/...
```

### Configuration
Run IRChuu~ for the first time and it will create a configuration file (you can also use `-data` and `-config` command-line arguments to specify a custom path):
```
$ irchuu
IRChuu! v0.6.0 (https://github.com/26000/irchuu)
2017/04/01 15:26:03 New configuration file was populated. Edit /home/26000/.config/irchuu.conf and run `irchuu` again!
```

Now edit the configuration file with your favourite editor (mine is `vim`, but I thought `nano` is more popular. Alternatively, you can just use a GUI editor like Kate):
```
$ nano ~/.config/irchuu.conf
```

The variables you *must* set are:
 - token, group in the `[telegram]` section
 - server, port, ssl, nick and channel in the `[irc]` section
 - probably serverpassword, password, sasl and chanpassword for IRC authentication

If you don't know where to get the Telegram token and groupname, refer to the next section.

Others are completely optional. The configuration file is well-documented, but if you have problems, feel free to open an issue on GitHub.

### Telegram bot setup
For IRChuu to work, you will need to create a Telegram bot as it works through the Telegram bot API. This is pretty simple:
1. Message [@botfather](http://t.me/botfather) inside Telegram. Send `/newbot` command.
2. It will ask some questions, answer all of them. You will have to think of a name and a nickname for your bot.
3. `@botfather` will send you a token. Insert it into your configuration file.
4. Type `/setprivacy` and choose your newly created bot nickname on the inline keyboard. Then choose **Disable**. This is important! If you forget to do it, messages from Telegram won't be relayed to IRC.
5. Optionally, type `/setuserpic` and upload a cute picure for your relay
6. Add your bot to your Telegram group (*Add members* and type the bot username there)
7. Launch IRChuu~ (just type `irchuu` in console once more)
8. Your bot will leave a message with the group id. That's totally ok, just copy the id into your config file, stop IRChuu (hit <kbd>Ctrl</kbd>+<kbd>C</kbd> in the terminal where IRChuu~ is running)
9. All set, now just launch IRChuu for the third time and enjoy!

### IRC setup
This one is easier. You can just insert your server and channel addresses into your configuration file and choose a nickname. If that nickname is already taken, IRChuu will think of a new one. If you want to own that nickname so that nobody takes it, register it and enter the password in the configuration file. Refer to your server's NickServ focumentation for details.

## Usage
Just type `irchuu`.

## Contributing
Feel free to fork this repo and make PRs. If you encounter a bug, please open an issue — that also helps! I will also be happy if you give IRChuu a star on GitHub.

## Special thanks
To [zephy!](https://github.com/zephyyy) and [Kotobank!](https://kotobank.ch/) for support and motivation.


changelog.txt
--- VERSION 1.1.1  / LAYER 19 ---
Fixed a bug which caused IRChuu~
to fail connecting to IRC. (You
don't have to upgrade if you don't
use TLS or don't experience this
issue.)

--- VERSION 1.1.0  / LAYER 18 ---
Added an option to wait before
joining a channel. Add the following
to the end of your config file to
use this feature:
```
# how many seconds to wait before joining the channel
joindelay = 0
```

Added a `status` command. It returns
IRC status in Telegram and vice
versa. In case it says "unable to
determine" in Telegram, add this to
the end of your config file and
adjust the timeout:
```
# how many seconds to wait for server to return the list of channels we're on
# (needed for /status command in Telegram; increase if it says you're not in
# channel when you are)
statustimeout = 2
```

IRChuu~ now will not wait for IRC
to disconnect and the database to
close before exiting if you press
Ctrl+C two times.

Fixed IRChuu~ not exiting on Ctrl+C
if it failed to join the channel.

Fixed a couple other bugs.

--- VERSION 1.0.0  / LAYER 17 ---
Now IRChuu~ will show you changelogs
if updates are available.

--- VERSION 0.10.1 / LAYER 16 ---
The Telegram bot no longer leaves
the group for which it isn't
configured. New Komf versions are
now supported.

--- VERSION 0.10.0 / LAYER 15 ---
IRC ignore lists are now available.
Add the following lines to the end
of your config file to use them:
```
# list of nicknames to ignore, i. e. messages by these users won't be relayed
ignorelist = ignoredbotnickname1,ignoredbotnickname2
```

Panic on exit was fixed. Errors
sending Telegram messages are now
reported.


config/config.go
package config

import (
	"html"
	"io/ioutil"
	"os"

	"gopkg.in/ini.v1"
)

const (
	// VERSION contains the IRChuu~ version.
	VERSION = "1.1.1"
	// LAYER contains IRChuu~ version in an integer (for comparison with
	// HQ server's last version).
	LAYER = 19
)

// ReadConfig reads the configuration file.
func ReadConfig(path string) (error, *Irc, *Telegram, *Irchuu) {
	cfg, err := ini.InsensitiveLoad(path)
	cfg.BlockMode = false

	tg, irc, irchuu := new(Telegram), new(Irc), new(Irchuu)

	err = cfg.Section("telegram").MapTo(tg)
	if err != nil {
		return err, irc, tg, irchuu
	}

	err = cfg.Section("irc").MapTo(irc)
	if err != nil {
		return err, irc, tg, irchuu
	}

	err = cfg.Section("irchuu").MapTo(irchuu)
	if err != nil {
		return err, irc, tg, irchuu
	}

	tg.Prefix = html.EscapeString(tg.Prefix)
	tg.Postfix = html.EscapeString(tg.Postfix)

	irc.IgnoreMap = map[string]bool{}
	for _, nickname := range irc.IgnoreList {
		irc.IgnoreMap[nickname] = true
	}

	if irc.StatusTimeout == 0 {
		irc.StatusTimeout = 2
	}

	return nil, irc, tg, irchuu
}

// PopulateConfig copies the sample config to <path>.
func PopulateConfig(file string) error {
	config := `# IRChuu configuration file. See https://github.com/26000/irchuu for help.
[irchuu]

# URI of your PostgreSQL database
# if blank, logging and kicking Telegram users from IRC will be unavailable
# (you will have to specify ?sslmode=disable if your database doesn't have TLS)
#
# examples:
# postgres://user:password@example.org:5432/database
# postgres://irchuu:irchuu@localhost/irchuu?sslmode=disable
dburi = 

# send usage statistics
# data what you will share:
# - the hashes of your Telegram group id and IRC channel
# - your IRChuu version
sendstats = true

# check for updates on each start
checkupdates = true

[telegram]
token = myToken
group = 7654321

# If message was sent more than <TTL> seconds ago, it won't be relayed
# 0 to disable
TTL = 300 # (seconds)

# prefix and postfix will be added before and after nicks
prefix = <
postfix = >

# allow sending messages without nick prefix (/bot command)
allowbots = true

# allow invites to the IRC channel from Telegram
allowinvites = false

# allow moderators in Telegram to kick users from IRC
# (bot needs to have permissions for that in IRC)
moderation = true

# download all media files to $XDG_DATA_HOME/irchuu or 
downloadmedia = false

# 'none', 'server' or 'pomf', where to store mediafiles from Telegram to show
# them in IRC
#
# 'server' will serve files over HTTP(S), needs 'serverport', 'baseurl',
# 'readtimeout' and 'writetimeout' to be set, 'downloadmedia' must be true
#
# 'pomf' will upload all media files to a pomf clone, needs 'pomf' to be set
#
# 'komf' will upload all media files to a komf (https://github.com/koto-bank/komf), needs 'komf' to be set
storage = none

## SERVER
# if certfilepath and keyfilepath are not nil, then will serve using HTTPS
certfilepath =
keyfilepath =

# request timeouts for the media server
readtimeout = 100 # (seconds)
writetimeout = 20 # (seconds)

# port for the media file server
serverport = 8080

# usually your protocol plus IP or domain plus the port, WITHOUT THE TRAILING SLASH
# don't forget to change http to https if enabled
baseurl = http://localhost:8080

## POMF
# the pomf clone url
# the following should work with irchuu:
# - https://p.fuwafuwa.moe
# - https://cocaine.ninja
# and many more. But some are retarded and won't.
pomf = https://p.fuwafuwa.moe

## KOMF
# a komf site url, you can set up your own: https://github.com/koto-bank/komf
komf =

# how much time will the file be stored for? (day, week, month)
komfdate = week

[irc]
server = irc.rizon.net
port = 6667
ssl = false
serverpassword =

nick = irchuu

# if not blank, will use NickServ to identify
password =

# if true, will use SASL instead of NickServ
sasl = false

# must be surrounded with backticks
channel = ` + "`" + `#irchuu` + "`" + `
chanpassword =
# how many seconds to wait before joining the channel
joindelay = 0

# colorize nicknames? (based on djb2)
colorize = true

# colors to be used, either codes or names
palette = 1,2,3,4,5,6,9,10,11,12,13

# prefix and postfix will be added before and after nicks
prefix = <
postfix = >

# maximum username length allowed, will be ellipsised if longer (0 to disable)
maxlength = 18

# lines in multi-line messages will be divided with this
# leave blank to send them as separate messages
ellipsis = "… "

# delay with which parts of multi-line message are sent to prevent anti-flood from kicking the bot
flooddelay = 500 # (milliseconds)

# allow ops in IRC to kick users from Telegram
# (bot needs to be a moderator in Telegram, also needs a database to be configured)
moderation = true

# who can kick users from the Telegram group:
# 1 — everybody, 2 — voices, 3 — halfops, 4 — ops, 5 — protected/admins, 6 — the owner
kickpermission = 4

# allow sending stickers from IRC? (by id)
allowstickers = true

# how often to poll the server for the users list
namesupdateinterval = 600 # (seconds)

# maximum number of messages sent on 'hist' command in IRC, works only with dburi set
maxhist = 40

# will send NOTICEs for private messages (help, hist, user count, etc) instead of PRIVMSGs
sendnotices = true

# forward join and part messages to Telegram
relayjoinsparts = true

# forward mode messages to Telegram
relaymodes = true

# rejoin automatically when kicked
kickrejoin = true

# announce the current topic to Telegram on join
announcetopic = true

# list of nicknames to ignore, i. e. messages by these users won't be relayed
ignorelist = ignoredbotnickname1,ignoredbotnickname2

# how many seconds to wait for server to return the list of channels we're on
# (needed for /status command in Telegram; increase if it says you're not in
# channel when you are)
statustimeout = 2
`
	return ioutil.WriteFile(file, []byte(config), os.FileMode(0600))
}

// Irchuu is the struct of common part in config.
type Irchuu struct {
	DBURI        string
	SendStats    bool
	CheckUpdates bool
}

// Irc is the stuct of IRC part in config.
type Irc struct {
	Server         string
	Port           uint16
	SSL            bool
	ServerPassword string

	Nick     string
	Password string
	SASL     bool

	Channel      string
	ChanPassword string
	JoinDelay    int

	Colorize      bool
	Palette       []string
	Prefix        string
	Postfix       string
	MaxLength     int
	Ellipsis      string
	FloodDelay    int
	AllowStickers bool

	Moderation          bool
	KickPermission      int
	MaxHist             int
	NamesUpdateInterval int
	SendNotices         bool
	RelayJoinsParts     bool
	RelayModes          bool
	KickRejoin          bool
	AnnounceTopic       bool

	IgnoreList []string
	IgnoreMap  map[string]bool

	StatusTimeout int

	Debug bool
}

// Telegram is the struct of Telegram part in config.
type Telegram struct {
	Token string
	Group int64

	TTL int64

	Prefix  string
	Postfix string

	AllowBots    bool
	AllowInvites bool
	Moderation   bool

	DownloadMedia bool
	Storage       string
	CertFilePath  string
	KeyFilePath   string
	ServerPort    uint16
	ReadTimeout   int
	WriteTimeout  int
	BaseURL       string
	DataDir       string
	Pomf          string
	Komf          string
	KomfDate      string
}

// muDeiPt5mAI8Ue==
